src/sync.h

```diff
@@ -185,21 +306,135 @@ class SCOPED_LOCKABLE CMutexLock
+#ifdef DEBUG_LOCKCONTENTION
+        }
+#endif
+// LogPrint("lck","ReadLock %p %s taken by %d\n", lock.mutex(), name ? name : "", boost::this_thread::get_id());
+#ifdef DEBUG_LOCKTIME
+        lockedTime = GetStopwatch();
+        if (lockedTime - startWait > LOCK_WARN_TIME)
+        {
+            LogPrint("lck", "Lock %s at %s:%d waited for %d ms\n", pszName, pszFile, nLine, (lockedTime - startWait));
+        }
+#endif
+    }
+
+    bool TryEnter(const char *pszName, const char *pszFile, int nLine)
+    {
+        name = pszName;
+        file = pszFile;
+        line = nLine;
+        EnterCritical(pszName, pszFile, nLine, (void *)(lock.mutex()), true);
+        if (!lock.try_lock())
+        {
+#ifdef DEBUG_LOCKTIME
+            lockedTime = 0;
+#endif
+            LeaveCritical();
+        }
+#ifdef DEBUG_LOCKTIME
+        else
+            lockedTime = GetStopwatch();
+#endif
+        return lock.owns_lock();
+    }
+
+public:
+    CMutexReadLock(Mutex &mutexIn, const char *pszName, const char *pszFile, int nLine, bool fTry = false)
+        SHARED_LOCK_FUNCTION(mutexIn)
+        : lock(mutexIn, boost::defer_lock)
+    {
+        if (fTry)
+            TryEnter(pszName, pszFile, nLine);
+        else
+            Enter(pszName, pszFile, nLine);
+    }
+
+    CMutexReadLock(Mutex *pmutexIn, const char *pszName, const char *pszFile, int nLine, bool fTry = false)
+        SHARED_LOCK_FUNCTION(pmutexIn)
+    {
+        if (!pmutexIn)
+            return;
+
+        lock = boost::shared_lock<Mutex>(*pmutexIn, boost::defer_lock);
+        if (fTry)
+            TryEnter(pszName, pszFile, nLine);
+        else
+            Enter(pszName, pszFile, nLine);
+    }
+
+    ~CMutexReadLock() UNLOCK_FUNCTION()
+    {
+        if (lock.owns_lock())
+        {
+            LeaveCritical();
+#ifdef DEBUG_LOCKTIME
+            int64_t doneTime = GetStopwatch();
+            if (doneTime - lockedTime > LOCK_WARN_TIME)
+            {
+                LogPrint(
+                    "lck", "Lock %s at %s:%d remained locked for %d ms\n", name, file, line, doneTime - lockedTime);
+            }
+#endif
+        }
+        // When lock is destructed it will release
+    }
+
+    operator bool() { return lock.owns_lock(); }
+};
+
+typedef CMutexReadLock<CSharedCriticalSection> CReadBlock;
+typedef CMutexLock<CSharedCriticalSection> CWriteBlock;
 typedef CMutexLock<CCriticalSection> CCriticalBlock;
 
-#define LOCK(cs) CCriticalBlock criticalblock(cs, #cs, __FILE__, __LINE__)
+#define READLOCK(cs) CReadBlock UNIQUIFY(readblock)(cs, #cs, __FILE__, __LINE__)
+#define WRITELOCK(cs) CWriteBlock UNIQUIFY(writeblock)(cs, #cs, __FILE__, __LINE__)
+#define READLOCK2(cs1, cs2) \
+    CReadBlock readblock1(cs1, #cs1, __FILE__, __LINE__), readblock2(cs2, #cs2, __FILE__, __LINE__)
+#define TRY_READ_LOCK(cs, name) CReadBlock name(cs, #cs, __FILE__, __LINE__, true)
+
+#define LOCK(cs) CCriticalBlock UNIQUIFY(criticalblock)(cs, #cs, __FILE__, __LINE__)
 #define LOCK2(cs1, cs2) \
```

Comment:
doesn't LOCK2 also need to be UNIQUIFY'd ?

Reply:
It would, but nobody is using 2 of them in the same scope. I'll fix in case it ever is used that way.

Reply:
fixed